Touchstone SAML

JPrevost opened the pull request:
## Status
**READY**

#### What does this PR do?

- changes auth from MIT Pilot to MIT Touchstone via in app SAML SP
- provides a dev mode auth option for PR builds and local work as
  registering a Touchstone SP for each PR and local environment is
  tedious
- also adds db migrate to release phase

#### Helpful background context (if appropriate)

https://mitlibraries.github.io/touchstone_saml.html

#### How can a reviewer manually see the effects of these changes?

Login at https://library-thesis-dropbox-staging.mit.edu

#### Todo:
- [x] Tests
- [x] Documentation
- [x] Stakeholder approval

#### Requires Database Migrations?
YES

#### Includes new or updated dependencies?
YES

thatandromeda commented on README.md:
```diff
@@ -7,6 +7,14 @@
+[docs/architecture-decisions directory](docs/architecture-decisions).
+
+[adr-tools](https://github.com/npryce/adr-tools) should allow easy creation of
+additional records with a standardize template.
```
s/standardize/standardized/;

JPrevost replied:
will fix

thatandromeda commented on .adr-dir:
```diff
@@ -0,0 +1 @@
+docs/architecture_decisions
```
Woohoo, documentation!

thatandromeda commented on app/controllers/application_controller.rb:
```diff
@@ -1,5 +1,6 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  skip_before_action :verify_authenticity_token
```
Wait, why?

JPrevost replied:
It's a form submission from the Touchstone system essentially so it does not have an authenticity token. It's from the docs.

thatandromeda commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -1,9 +1,19 @@
       flash[:notice] = "Welcome #{@user.email}!"
     end
+
+    def developer
```
This could stand to have a comment (and also the docs should be very clear about ensuring `FAKE_AUTH_ENABLED` is false in production).

thatandromeda replied:
Super handy, though. I see why it's here.

JPrevost replied:
Cool. I can doc that up. Readme for sure and a note here.

thatandromeda commented:
Yay, progress!

This requires several environment variables to be set - can those variables and their desired settings be documented somewhere?

I am torn on the developer authentication. I see how this is super useful for testing purposes, but it is also very explicitly a thing that makes everyone an admin user, which is potentially a security-hole gun sitting around so that we can shoot ourselves with it. I would at the least like to see that function liberally festooned with documentation as to how _not_ to shoot ourselves with it, but if you have additional ideas for making it hard to trip over by accident that would be great too.

JPrevost commented:
Yeah. We need dev mode for PR builds as well as those URLs aren't registered with Touchstone but people will sometimes need to login to test stuff. I'll add some new docs and warnings around it as well as open a ticket to consider preventing it from running on the production heroku instances... but that gets a bit tricky as prod staging and pr builds all run with `production` as the environment flag. We can do better though with some focused thought I bet.

JPrevost commented:
@thatandromeda I've added notes on the ENV and added an initial warning. I will be opening a new ticket to further prevent enabling dev auth on staging / production servers.

JPrevost commented:
And the promised ticket to safety it up: https://mitlibraries.atlassian.net/browse/TI-23

thatandromeda commented on README.md:
```diff
@@ -13,14 +13,50 @@ This repository contains Architecture Decision Records in the
+
+For SAML authentication, you will need all of the following.
+
+[DLE Docs on SAML](https://mitlibraries.github.io/touchstone_saml.html)
```
🌈 for these existing.
